keps/sig-cloud-provider/1959-service-lb-class-annotation/README.md

```diff
@@ -0,0 +1,185 @@
+* I MUST use a hardware-based loadbalancer for certain applications due to specific protocols
+only available there.
+* I want to use the cloud provider's default load balancer for any applications that do not
+rely on protocols from hardware load balancers.
+
+### Risks and Mitigations
+
+Many Service Type=LoadBalancer implementations today support a lot of knobs via annotations already.
+Introducing yet another annotation for Service Type=LoadBalancer is not ideal, but this is better than
+every cloud provider supporting their own "skip this Service" annotation.
+
+## Design Details
+
+Introduce a new Service annotation `service.kubernetes.io/load-balancer-class`.
+
+If the loadbalancer class annotation is not set, the existing cloud provider
+will assume ownership of the Service Type=LoadBalancer resource. This is required
+to not break existing clusters that assume Service Type=LoadBalancer is always
+managed by the cloud provider.
+
+Required updates to service controller:
+* if the class annotation is NOT set for a Service, allow the cloud provider
+to reconcile the load balancer.
+* if the class annotation IS set for a Service, skip reconciliation of the Service
+by the cloud provider.
+
+### Test Plan
+
+Unit tests:
+* test that service controller does not call the cloud provider if the class annotation is set.
+* the annotation `service.kubernetes.io/load-balancer-class` is not accepted when the feature gate `ServiceLoadBalancerClass` is disabled.
+
+E2E tests:
+* test that creating a Service with an unknown class annotation results in no load balancer being created for a Service.
+
+### Graduation Criteria
+
+N/A since we can't apply alpha/beta/GA criteria for annotations.
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, use of this annotation will be allowed. On downgrade, service controller
+may ignore existing Services with the annotation, leading to multiple implementations
+trying to create load balancers. On downgrade, if the class annotation is used
+and there are multiple implementations of Service Type=LoadBalancer, a user must ensure
+there is only 1 implementation of Service Type=LoadBalancer in the cluster.
+
+Though the downgrade scenario isn't ideal, it is assumed if that a cluster was upgraded to v1.20,
+and already has multiple Service Type=LoadBalancer implementations enabled, it will likely not be
+downgrading to v1.19 anytime soon.
+
+### Version Skew Strategy
+
+N/A since this only impacts one component.
+
+## Implementation History
+
+- the `Summary`, `Motivation`, `Proposal` and `Design Details` sections was merged, signaling SIG acceptance
+
+## Drawbacks
+
+* Annotations are a clunky way to implement "Class" semantics to Service Type=LoadBalancer.
+* In **most** clusters, a single Service Type=LoadBalancer implementation from the cloud provider is sufficient.
+* The potential risks during downgrade can cause outages if Service Status is updated by the wrong load balancer implementation.
+
+## Alternatives
```

Comment:
Another alternative is a field.  Not that I want a new field, but it's not inconsistent with past decisions AND it has established precedent for introducing anew.

Add it as an alpha, gated, next release ungate.

The questions then are whether it's really worth a field or not, what the contents would be, and will this be a slippery slope to parameter blocks?

Should existing provider-specific annotations switch to this?

Reply:
> The questions then are whether it's really worth a field or not, what the contents would be, and will this be a slippery slope to parameter blocks?

Personally I don't think a field make sense here since the behavior we're introducing is to toggle an implementation of the API and nothing about the API itself. There's also some precedent here for using annotations, it's just a matter of consolidating all implementations on a single annotation as opposed to a custom annotation for every cloud provider.

> Should existing provider-specific annotations switch to this?

In theory yes, but in practice I think most cloud providers will end up supporting both the existing provider-specific annotation and the generic one. For the cloud providers that have not introduced provider-specific annotations yet, they would get this for free.

Reply:
We have other LB policy fields, so this isn't out of line with precedent.  Additionally, "class" is a field in all of the other lowest-level APIs (storage, ingress, gateway, runtime...) so again, this is in-line.  The main advantage of a field is that it is the well-paved path.  We know how to introduce fields.

We *do* need to define what this field holds - arbitrary strings?  Label-style names (similar to storage provisioners)?  Something else?

Reply:
I guess I am saying OK. I'm not super excited about needing this, but I ACK the use-case.

Reply:
Okay I think I'm convinced to use fields, I'll update the PR soon.